aws_lambda_builders/workflows/java_gradle/DESIGN.md

```diff
@@ -0,0 +1,117 @@
+be run on an older VM if configured using the `targetCompatibility` and
+`sourceCompatibility` properties of the Java plugin. Therefore, it is not
+sufficient to check the version of the local JDK, nor is it possible to check
+the value set for `targetCompatibility` or `sourceCompatibility` since it can
+be local to the compile/build task. At best, we can check if the local
+version of the JDK is newer than Java 8 and emit a warning that the built
+artifact may not run in Lambda.
+
+Gradle projects are configured using `build.gradle` build scripts. These are
+executable files authored in either Groovy or since 5.0, Kotlin, and using the
+Gradle DSL. This presents a similar problem to `setup.py` in the Python world in
+that arbitrary logic can be executed during build time that could affect both
+how the customer's artifact is built, and which dependencies are chosen.
+
+To solve the issue where the build environment could affect the `build.gradle`
+script, a `--use-container` option will be provided, to give the option of
+building in an environment that closely mimics the Lambda environment.
+
+## Implementation
+
+We leverage Gradle to do all the heavy lifting for executing the `build.gradle`
+script which will resolve and download the dependencies and build the project.
+To create the distribution ZIP, we insert a new Gradle `task` that packages the
+compiled classes, resouces, and dependencies into a ZIP.
+
+### sam build
+
+#### Step 1: Check Java version and emit warning
+
+Check whether the local JDK version is <= Java 8, and if it is not, emit a
+warning that the built artifact may not run in Lambda unless a) the project is
+properly configured (i.e. using `targetCompatibility`) or b) the project is
+build within the Lambda-compatibile container.
+
+#### Step 2: Synthesize a new build script
+
+There is no standard task in Gradle to create a distribution ZIP (or uber JAR).
+We must synthesize a new build script that we will run to actually produce the
+Lambda artifact. This is so we can insert a new task that will assemble the
+distribution ZIP. The new script is identical to the customer's original
+`build.gradle` except that it imports our own `sam-ext.gradle` file provided by
+this package.
+
+It will do something similar to:
+
+```sh
+cp build.gradle build.gradle.sam-build
+
+cp /path/to/sam-ext.gradle /tmp
+
+echo 'apply from: "/tmp/sam-ext.gradle"' >> build.gradle.sam-build
+```
+
+where the contents of `sam-ext.gradle` contains the definition of the
+`samBuildZip` task:
+
+```gradle
+// Include the project classes and resources in the root, and the dependencies
+// under lib
+task samBuildZip(type: Zip) {
+    depends on build
+    from sourceSets.main.output
+    dependsOn configurations.runtimeClasspath
+    into('lib') {
+        from configurations.runtimeClasspath
+    }
+}
+```
+
+#### Step 3: Build and package
+
+```sh
+gradle -b build.gradle.sam-build samBuildZip
```

Comment:
Do we want to use gradlew? So that we don't have to worry about Gradle versions having to match.

Reply:
That would necessitate a preexisting `gradlew` file though right? Maybe we can look for that first and fallback to a local `gradle` executable if we can't find it.

Reply:
It's standard practice to check the gradlew system into Git

Reply:
But for last ditch we could search the path

Reply:
Right but they may not have generated one right (even if the majority of projects do)?

Reply:
Ya, gradlew should have precedence and fall back to search if you need to.

Reply:
Not sure I understand this search the path piece here. Why do we need to search for anything? Why wouldn't the gradlew file be provided in the `source_dir`?